mmv1/third_party/cai2hcl/common/utils.go

```diff
@@ -63,3 +75,54 @@ func hashicorpCtyTypeToZclconfCtyType(t hashicorpcty.Type) (cty.Type, error) {
 	}
 	return ret, nil
 }
+
+func NewConfig() *transport_tpg.Config {
+	return &transport_tpg.Config{}
+}
+
+// normalizeNodeRec traverses the output map recursively, removes fields which are
+// not a part of TF schema and converts unmarshallable "schema.Set" objects to arrays.
+func normalizeNodeRec(node interface{}, mapSchema map[string]*schema.Schema) interface{} {
+	switch node.(type) {
+	case map[string]interface{}:
+		mapObj := node.(map[string]interface{})
+		newMapObj := map[string]interface{}{}
+
+		if mapSchema == nil {
+			// mapSchema is applicable only for key-value objects.
+			return node
+		}
+
+		for key, value := range mapObj {
+			propertySchema := mapSchema[key]
+			if propertySchema == nil {
+				// Strip unknown fields.
```

Comment:
I have elaborated in the comment, why it makes sense to not fail conversion in this case.
The real life scenario is that GCP Console UI implements new fields in parallel with the API, maybe even before the TF support (i.e. during internal or external(staging/alpha) testing of the new field).
This does not necessitate the need of this code (lack of TF support of the feature is another case than that), but improves the chance of version skews and, I think, requires us to return "partial" responses whenever possible.

In the close future we need to store the errors and return them to the caller, so that UI users see that the provided TF is incomplete / has errors.

Reply:
Hm, actually this all is not needed once I have redone recursion via schema, thanks for the tip!

Reply:
Just in case it's important, I wanted to clarify what I was trying to say. I understand that there could be skew between the API and TGC, where there are new fields in the API but not yet in Terraform. However, that data should already be cleaned by the time we get to `MapToCtyValWithSchema`, because the flatten functions convert the API object to a `map[string]interface{}` with an explicit set of fields, which are all present in the provider. This is already a situation that happens with the Terraform providers - the data stored in state is also the result of calling the resource's flatteners on the API response, and only contains fields that are supported in Terraform.

The conversion to cty values is because that's what we need for the hclwrite, but the data should already only contain Terraform-supported fields by this point anyway.

Reply:
Hm, I thought of the version skew when:
1. New field has been added to MMv1 YAML (assuming cai2hcl converters are auto-generated)
2. CI propagates changes to TPG version X1
3. CI propagated changes to TGC version Y1, which still depends on TPG lower than X1

AFAIU, at this point version conflict happens in TGC Y1.
Also when imported inside Google repo, TPG version can become older (or newer).

Reply:
hmm, yeah, if you forced TGC to compile with an older version of TPGB that doesn't support a particular field, then I suppose you could end up with this scenario. In general TGC versions should align 1:1 with TPG(B) versions but it's also fine to handle the case where they don't for some reason.